### PR TITLE
Safety check low_cpu_mem_usage when in 4bit or 8bit

### LINKED ISSUE #1858: ValueError: weight is on the meta device, we need a value to put in on device with 4bit
### System Info

```Shell
- `Accelerate` version: 0.21.0
- Platform: Linux-5.14.0-284.25.1.el9_2.x86_64-x86_64-with-glibc2.34
- Python version: 3.11.4
- Numpy version: 1.24.3
- PyTorch version (GPU?): 2.0.1+cu117 (True)
- PyTorch XPU available: False
- PyTorch NPU available: False
- System RAM: 501.25 GB
- GPU type: NVIDIA GeForce RTX 3090
```

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] One of the scripts in the examples/ folder of Accelerate or an officially supported `no_trainer` script in the `examples` folder of the `transformers` repo (such as `run_no_trainer_glue.py`)
- [X] My own task or dataset (give details below)

### Reproduction

Related to [this issue](https://github.com/huggingface/accelerate/issues/1289) but I was asked to open a new issue. I can't show you the strack tace now (all GPUs are busy, can't re-run code) but it is the same as the linked issue except that for me it occurred on CUDA. It goes away when enabling `--low_cpu_mem_usage`.

I don't have a **minimal** example, but full code that I used that triggered this is [here](https://github.com/BramVanroy/llm-finetuning/blob/main/src/llm_finetuning/run_chat_modeling.py). 

Use this config as `config.json`

```json
{
  "log_on_each_node": false,
  "report_to": "wandb",

  "per_device_train_batch_size": 2,
  "per_device_eval_batch_size": 2,
  "gradient_accumulation_steps": 8,

  "optim": "paged_adamw_32bit",
  "learning_rate": 2e-4,
  "weight_decay": 0.01,
  "lr_scheduler_type": "cosine",
  "warmup_ratio": 0.03,
  "adam_beta1": 0.9,
  "adam_beta2": 0.95,

  "save_steps": 20,
  "evaluation_strategy": "steps",
  "eval_steps": 20,
  "logging_first_step": true,
  "logging_steps": 1,
  "num_train_epochs": 2,

  "early_stopping_patience": null,
  "early_stopping_threshold": null,

  "dataset_name": "BramVanroy/dutch_chat_datasets",
  "group_by_length": true,
  "preprocessing_num_workers": 24,
  "dataset_batch_size": 1000,
  "output_dir": "results",
  "validation_split_percentage": 5,
  "load_best_model_at_end": true,
  "save_total_limit": 3,

  "trust_remote_code": true,
  "model_name_or_path": "BramVanroy/llama2-13b-ft-mc4_nl_cleaned_tiny",
  "max_length": 4096,
  "use_nested_quant": true,
  "bf16": true,
  "tf32": true,
  "load_in_4bit": true,
  "bnb_4bit_quant_type": "nf4",
  "torch_dtype": "bfloat16",
  "bnb_4bit_compute_dtype": "bfloat16",
  "lora_alpha": 32,
  "lora_r": 4,
  "low_cpu_mem_usage": false,
  "ddp_find_unused_parameters": true,
  "use_peft": true,
  "max_grad_norm": 1.0,
  "use_flash_attention": true,

  "do_train": true,
  "do_eval": true
}
```

Call as you would

```shell
python run_chat_modeling.py config.json
```

In my case, I used deepspeed on 4x 3090s (the config is in the repository that I linked):

```shell
deepspeed src/llm_finetuning/run_chat_modeling.py chat_config.json --deepspeed deepspeed_configs/ds_config_zero2.json
```

Hope that helps to reproduce the issue?

### Expected behavior

Either a fix (if it is a bug) _or_ a better error message if it is expected behavior.

It goes away when enabling `--low_cpu_mem_usage`. @sgugger mentioned that loading in 4bit must require `low_cpu_mem_usage=True`. If that is the case, then the solution should be relatively straightforward as a better error message.

### PR DESCRIPTION
**Pull request by BramVanroy**
Currently, the check is only done for `is None` after which `low_cpu_mem_usage ` is set to True. But that can lead to unexpected behavior, especially if it is set to False. The result would be a hard-to-debug error trace (cf. related issue). This PR makes sure that when in 4bit or 8bit, `low_cpu_mem_usage` is set to True regardless of its current value and it will log a warning to the user to notify them of the change.

closes https://github.com/huggingface/accelerate/issues/1858

### CODE REVIEW

**Comment by ArthurZucker** [comment]
cc @younesbelkada

**Review by younesbelkada — APPROVED** [review]
Sounds good, thank you @BramVanroy !

File: src/transformers/modeling_utils.py [hunk 1]
```diff
@@ -2478,8 +2478,10 @@ def from_pretrained(
                     "Setting device_map to {'':torch.cuda.current_device()}."
                     "If you want to use the model for inference, please set device_map ='auto' "
                 )
-                if low_cpu_mem_usage is None:
-                    low_cpu_mem_usage = True
+
+            if low_cpu_mem_usage is not True:
+                low_cpu_mem_usage = True
+                logger.info("Changed `low_cpu_mem_usage` to `True` as required in 4bit and 8bit")
```
Comment (ArthurZucker) [inline src/transformers/modeling_utils.py:2484]:
```suggestion
                logger.warning("Changed `low_cpu_mem_usage` from `False` to  `True` as it is required to load in 4bit and 8bit")
```
let's rather warn !

**Review by ArthurZucker — APPROVED** [review]
Thanks for the fix, can we `warn` rather than put is as an info please

**Comment by ArthurZucker** [comment]
Running `make style` should fix the CI 😉

**Pull request closed without merging**